### DISCUSSION TITLE
Context params

### DISCUSSION THREAD

**Discussion by dandeduck** [oakserver/oak#369] [Q&A]
I wanted to receive `ctx : Context` on my controllers, but it turns out that it lacks some fields compared to `any`, such as `ctx.params`. I saw a similar issue where the issuer was told to use `ctx.request.url.searchParams` but it was always empty, even when `ctx.params` wasn't. Is there a way to get it without using the `any` type? I assume that I am missing something... 

If you want to take a look at the source code, it's in this project: https://github.com/dandeduck/weebify-leaderboard under the development branch

**Comment by kitsonk** [comment]
There is no code in that repo.

`ctx.params` is specifically a [`RouterContext`](https://doc.deno.land/https/deno.land/x/oak/mod.ts#RouterContext) "thing" and doesn't exist on [`Context`](https://doc.deno.land/https/deno.land/x/oak/mod.ts#Context).  It will only exist in [`RouterMiddleware`](https://doc.deno.land/https/deno.land/x/oak/mod.ts#RouterMiddleware).

**Reply by dandeduck** [reply]
>  under the development branch

There is code, but anyway... It was hard to know that `RouterContext` is a things, since even when printing it (under any), it says `Context`. Thanks for the answer!

**Reply by kitsonk** [reply]
Because it is only an interface, it doesn't exist at runtime. If you add [RouterMiddleware](https://doc.deno.land/https/deno.land/x/oak/mod.ts#RouterMiddleware) to an instance of the [Router](https://doc.deno.land/https/deno.land/x/oak/mod.ts#Router), you will noticed that the context is contually inferred to be `RouterContext` and will try to infer the generics around the route params and state.

If you are trying to create standalone middleware you will then add to a router, best to declare it is something like this:

```ts
import type { RouterMiddleware, RouteParams, State } from "https://deno.land/x/oak@v7.7.0/mod.ts";

interface MyParams extends RouteParams {
  // ...
}

interface MyState extends State {
  // ...
}

const myMiddleWare: RouterMiddleware<MyParams, MyState> = (ctx, next) => {
  // ctx and next will be typed correctly
};
```
